### ISSUE TITLE
Listening to events triggered after the app is launch using Firebase Dynamic Links.

### ISSUE THREAD

**Issue by ankit18singh** [chemerisuk/cordova-plugin-firebase-dynamiclinks#54]
Basically, 
Once the app is launched from the Firebase Dynamic Links, we do receive the `link` parameter by subscribing to the `onDynamicLink` method, but we have certain scenarios which are not documented where we need to track some additional data and proceed accordingly, which are mentioned below:

-  How can we differentiate and track if the app is launched normally or using the generated Firebase Dynamic Link?
-  What will happen if we directly click on the URL configured with the `APP_DOMAIN` variable?

So in the current version, do we have any way of listening to events which get triggered once the app is launched using generated dynamic link and map the navigation parameters passed within to navigate the app to the requested view?

Please guide me to a better approach or help redirect me to a service which can help me meet these requirements.

Thanks.

**Comment by ankit18singh** [comment]
The above description has been updated.

**Comment by sagrawal31** [comment]
@chemerisuk here at [FirebaseDynamicLinksPlugin.java#L81](https://github.com/chemerisuk/cordova-plugin-firebase-dynamiclinks/blob/v4.0.1/src/android/FirebaseDynamicLinksPlugin.java#L81), why are we using `.continueWith` instead of [addOnSuccessListener](https://firebase.google.com/docs/reference/android/com/google/firebase/dynamiclinks/FirebaseDynamicLinks)?

Because, as of now, if app is launched directly or through the `APP_DOMAIN` deeplink, then `cordova.plugins.firebase.dynamiclinks.onDynamicLink` is never executed.

**Comment by sagrawal31** [comment]
So that can invoke the `onDynamicLink` JS callback with a key that says there was a pending dynamic link or not.

**Comment by sagrawal31** [comment]
@chemerisuk your thoughts!

**Comment by chemerisuk** [comment]
@ankit18singh ask google, I'm not developing the API, the plugin just returns data provided by firebase SDK.

@sagrawal31 `.continueWith` is used to simplify code. Not sure I understood your use case.

**Comment by sagrawal31** [comment]
Hi @chemerisuk apologies for the late reply. Let me rephrase my comment/question. I have this code written:

```
cordova.plugins.firebase.dynamiclinks.onDynamicLink(function(data) {
    console.log("Dynamic link click with data:", data);
});
```

Now, the app is launched in 3 different scenarios:

Scenario | Above callback executes
----|-----
The app is freshly launched directly from the app icon | No
The app is launched from `APP_DOMAIN` URL | No
The app is launched from deep link URL `PAGE_LINK_DOMAIN` | Yes

Consider, we want to hit our backend server on the app launch and we wants to send some data received from the dynamic link. So in the first two scenarios, we can't get to know that the app is launched from the dynamic link or not.

As an alternative, we had to put a 10-second wait and we check if the callback is not executed within 10 seconds that means the app was launched either from the first two scenarios.

So how we are supposed to get the data callback (`true`/`false`) from this plugin that if we do have any pending dynamic link data?

I hope I have explained the problem properly this time 😄

**Comment by numerized** [comment]
I have the same problem, this is a bummer, we have to add a timer while app is launched from home app icon in case we receive a dynamiclink. There should have a way to know beforehand if the app has been launched without dynamic links.

Is there a way ? I cannot find it yet

**Comment by sagrawal31** [comment]
No @numerized, author does not want to implement the change for some reason as commented in the PR https://github.com/chemerisuk/cordova-plugin-firebase-dynamiclinks/pull/57#issuecomment-565487519

**Comment by numerized** [comment]
That's really a pitty, I have a timeout function to detect this, is there a better way to solve this issue, I see your PR is only covering Android ?

**Comment by sagrawal31** [comment]
Yes, I have limited knowledge of iOS.

I used a double callback to the server in case we receive deep link from this plugin, that was my solution. I implemented timeout concept at first but that's not reliable

**Comment by chemerisuk** [comment]
@numerized @sagrawal31 guys, I understand you requirement. My proposal is to introduce a new method `getDynamicLink` that will read current dynamic link data or return null if it does not exist. In that case you can call it on start. Also to change behavior for callback `onDynamicLink` so it won't fire on app start. What do you think?

**Comment by sagrawal31** [comment]
Hi @chemerisuk thanks for jumping in and reconsidering the issue.

> My proposal is to introduce a new method `getDynamicLink` that will read current dynamic link data or return null if it does not exist.

Looking good to me. This one can be done as a one-time method to get the dynamic link.

> Also to change behavior for callback onDynamicLink so it won't fire on app start.

Didn't get this part. Do you mean that `onDynamicLink` won't be called on the first time (on app launch) but the subsequent calls (like if the app is already opened and someone again open it from a dynamic link) will return result in `onDynamicLink`? (looks like `getToken()` and `onTokenRefresh()` methods for firebase messaging).

If this is the case, I agree with you.

**Comment by chemerisuk** [comment]
@sagrawal31 right, very similar to what firebase messaging API does. To illustrate check the code example below:
```js
cordova.plugins.firebase.dynamiclinks.getDynamicLink().then(function(data) {
    if (data) {
        console.log("Read dynamic link data on app start:", data);
    } else {
        console.log("App wasn't started from a dynamic link");
    }
});
cordova.plugins.firebase.dynamiclinks.onDynamicLink(function(data) {
    console.log("Dynamic link click with data:", data);
});
```

**Comment by chemerisuk** [comment]
Just pushed code changes at https://github.com/chemerisuk/cordova-plugin-firebase-dynamiclinks/commit/d6b5246390baed870b1a972b98836ba126835c40.

@numerized @sagrawal31  Do you mind to test the new method from master branch?

**Comment by sagrawal31** [comment]
@chemerisuk I'll test the change by Friday, if that works

**Comment by mcfarljw** [comment]
@chemerisuk it looks like an import is missing. Here is the error when running on Android:

```
android/app/src/main/java/by/chemerisuk/cordova/firebase/FirebaseDynamicLinksPlugin.java:103: error: cannot find symbol
                .addOnFailureListener(new OnFailureListener() {

symbol:   class OnFailureListener
location: class FirebaseDynamicLinksPlugin
```

It looks like adding `import com.google.android.gms.tasks.OnFailureListener;` should fix that up.

**Comment by mcfarljw** [comment]
@chemerisuk I tested this on both Android and iOS. `getDynamicLink()` did return a promise containing the correct data on app start. `onDynamicLink` fires a callback when a dynamic link is clicked when the app is already open. I did have to make some changes which I opened pull requests for and referenced here to get it to build though.

**Comment by chemerisuk** [comment]
@mcfarljw great, thanks for valuable input! I did merge bug fixes you've found. Let's resolve the last PR and I think we can publish a new release.

**Comment by numerized** [comment]
big thank to you 3!
that's a real improvement.

**Comment by chemerisuk** [comment]
New method published in v4.7.0

**Comment by sagrawal31** [comment]
Thanks, @chemerisuk for your time.

**Comment by numerized** [comment]
It works!
Thanks a million, app is more snappy now
👍
